src/main/java/org/mockito/internal/matchers/CapturingMatcher.java

```diff
@@ -26,16 +28,17 @@ public String toString() {
     }
 
     public T getLastValue() {
-        if (arguments.isEmpty()) {
-            throw noArgumentValueWasCaptured();
-        }
-
-        return (T) arguments.getLast();
+        synchronized (arguments) {
+            if (arguments.isEmpty()) {
+                throw noArgumentValueWasCaptured();
+            }
 
+            return (T) arguments.get(arguments.size() - 1);
+        }
     }
 
     public List<T> getAllValues() {
-        return (List) arguments;
+        return Arrays.asList((T[]) arguments.toArray());
```

Comment:
@SimY4 
Note that you must wrap this code in an synchronized block. Otherwise there is sitll room for a race condition when an argument is added in `captureFrom(..)`.